02-intermediate-ruby/advanced-variables.md

```diff
@@ -15,137 +15,296 @@
   end
+
 end
+
+library1 = Library.new("Woodland Library")
+library2 = Library.new("Northview Library")
+
+puts library1.library_name()
+puts library2.library_name()
 ```
 
+This code outputs the following:
+```
+Woodland Library
+Northview Library
+```
+
+In this example, the different instances of Library have different values for `library_name`.
+
+#### Exercise
+To build on this, try defining more parameters and instance variables for the `Library` class that would vary instance by instance, like open hours or featured book.
+
+As we learn about some of the more broad variables below, remember that we can use local and instance variables to avoid some of the issues we are about to discuss.
+
 ### Class Variables
 - Class variables begin with `@@`
 - Class variables are available to the entire class (in any method)
 - Class variables will raise an error if they are read before they're created
-- Class variables can cause problem down the road (**avoid using them**)
+- Class variables can cause problems later (**avoid using them**)
 - Class variables are sometimes used for application configuration
 
 ```ruby
-class Ghost
-  @@ghosts_in_the_world = 0
-  attr_reader :name
+class Library
+  attr_reader :library_name
+  @@tagline = "Welcome to all of the libraries:"
+
+  def initialize(name)
+    @library_name = name
+    @@tagline = @@tagline + " " + @library_name
+  end
+
+  def tagline
+    puts @@tagline
+  end
+
+end
+
+library1 = Library.new("Woodland Library")
+library2 = Library.new("Northview Library")
+library1.tagline()
+library2.tagline()
+```
+
+This code outputs the following:
+```
+Welcome to all of the libraries: Woodland Library Northview Library
+Welcome to all of the libraries: Woodland Library Northview Library
+```
+
+Here, we observe that all `Library`s can access and change the same `@@tagline` variable. This is useful in some cases, but can be confusing and obscure in most cases.
+
+Let's explore one more example. Imagine a world where if one library opens, every library is open, and if one library closes, every library is closed. We'll model it like so:
+
+```ruby
+class Library
+  @@status = "closed"
+
+  def status
+    puts "The library is " + @@status
+  end
+
+  def open_library
+    @@status = "open"
+  end
+
+  def close_library
+    @@status = "closed"
+  end
+end
+
+library1 = Library.new
+library2 = Library.new
+library1.status()
+library2.status()
+library1.open_library()
+library1.status()
+library2.status()
+library2.close_library()
+library1.status()
+library2.status()
+```
 
-  def initialize(name_var)
-    @@ghosts_in_the_world +=1
-    @name = name_var
+This code outputs the following:
+```
+The library is closed
+The library is closed
+The library is open
+The library is open
+The library is closed
+The library is closed
+```
+
+Here we see our desired behavior: if one `Library` opens, all `Library`s open, and if one `Library` closes, all `Library`s close.
+
+#### What about Subclasses?
```

Comment:
Since we don't talk about subclasses until later in the curriculum, it might make sense to pull this out into a separate resource for when we talk specifically about using subclasses